src/consensus/vote.cpp

```diff
@@ -570,6 +614,82 @@ std::string VoteManager::getJsonStr(std::vector<Vote> const& votes) {
   return ptroot.toStyledString();
 }
 
+std::vector<std::shared_ptr<Vote>> VoteManager::getProposalVotes(uint64_t pbft_round) {
+  std::vector<std::shared_ptr<Vote>> proposal_votes;
+
+  sharedLock_ lock(verified_votes_access_);
+  for (auto const& voted_value : verified_votes_[pbft_round][1]) {
+    for (auto const& v : voted_value.second) {
+      proposal_votes.emplace_back(std::make_shared<Vote>(v.second));
+    }
+  }
+
+  return proposal_votes;
+}
+
+VotesBundle VoteManager::getVotesBundleByRoundAndStep(uint64_t round, size_t step, size_t two_t_plus_one) {
+  std::vector<Vote> votes;
+
+  sharedLock_ lock(verified_votes_access_);
+  auto found_round_it = verified_votes_.find(round);
+  if (found_round_it != verified_votes_.end()) {
+    auto found_step_it = found_round_it->second.find(step);
+    if (found_step_it != found_round_it->second.end()) {
+      for (auto const& voted_value : found_step_it->second) {
+        if (voted_value.second.size() >= two_t_plus_one) {
+          auto voted_block_hash = voted_value.first;
+          votes.reserve(two_t_plus_one);
```

Comment:
I already asked but this reserve was somewhere else so for reminding the conversation I asked:

I guess this is some mistake. Why would you reserve always for the same size inside for loop ?

@gaoqi8649 gaoqi8649 6 days ago Author Member
Because only need 2t+1 number of votes to make consensus

So again it does not make any sense to reserve vector size in loop. What are you trying to achieve by this ?

Reply:
This will only reserve one time and will return VotesBundle(true, voted_block_hash, votes)